bls/bls.go

```diff
@@ -1,79 +1,110 @@
 package bls
 
 import (
-	"crypto/rand"
 	"errors"
+	"math/big"
 
-	bls12381 "github.com/kilic/bls12-381"
+	bls12381 "github.com/consensys/gnark-crypto/ecc/bls12-381"
+	"github.com/consensys/gnark-crypto/ecc/bls12-381/fr"
 )
 
-// Heavily inspired by:
-// https://github.com/protolambda/bls12-381-util/blob/master/signatures.go
-// Thank you for the excellent code.
-
-var domain = []byte("BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_POP_")
-
 const (
-	PublicKeyLength int = 48
-	SecretKeyLength int = 32
-	SignatureLength int = 96
+	PublicKeyLength = bls12381.SizeOfG1AffineCompressed
+	SecretKeyLength = fr.Bytes
+	SignatureLength = bls12381.SizeOfG2AffineCompressed
 )
 
 type (
-	PublicKey = bls12381.PointG1
-	SecretKey = bls12381.Fr
-	Signature = bls12381.PointG2
+	PublicKey = bls12381.G1Affine
+	SecretKey = fr.Element
+	Signature = bls12381.G2Affine
 )
 
 var (
+	_, _, g1One, _            = bls12381.Generators()
+	domain                    = []byte("BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_POP_")
 	ErrInvalidPubkeyLength    = errors.New("invalid public key length")
 	ErrInvalidSecretKeyLength = errors.New("invalid secret key length")
 	ErrInvalidSignatureLength = errors.New("invalid signature length")
 	ErrSecretKeyIsZero        = errors.New("invalid secret key is zero")
+	ErrPointNotOnCurve        = errors.New("point not on the curve")
+	ErrPointNotInSubGroup     = errors.New("point not in correct subgroup")
 )
 
 func PublicKeyToBytes(pk *PublicKey) []byte {
-	return bls12381.NewG1().ToCompressed(pk)
+	pkBytes := pk.Bytes()
+	return pkBytes[:]
 }
 
 func SecretKeyToBytes(sk *SecretKey) []byte {
-	return sk.ToBytes()
+	skBytes := sk.Bytes()
```

Comment:
This is correct though one thing to check is endianess -- gnark serialized `fr.Element` using big endian in the past, now you can specify it -- if there are tests for this function then they would fail if endianness was wrong anyways

Reply:
Oh, good idea, I will check that.